src/main/java/cs361/battleships/models/Board.java

```diff
@@ -48,27 +49,52 @@ public Result attack(int x, char y) {
 		//Initialize result with status TBD
 		Result res = new Result();
 		res.setLocation(sq1);
-		//If ship is present on coordinates
 
-		if(isDuplicateAttack(sq1) && !isCaptainsQuarter(x,y) || (isCaptainsQuarter(x,y) && !(this.getShip(x,y) == null) && this.getShip(x,y).isSunk()) ) {
+		//If ship is present on coordinates
+		//repeat attacks only during laser phase or on unsunk captain's quarters during bomb phase
+		if(!laserEnabled && isDuplicateAttack(sq1) && shipPresent(x,y)) {
+			for (Ship ships : getShip(x, y)) {
+				if (!isCaptainsQuarter(x, y, ships) || isCaptainsQuarter(x, y, ships) && ships.isSunk()) {
+					res.setResult(AttackStatus.INVALID);
+					return res;
+				}
+			}
+		} else if (isDuplicateAttack(sq1) && !laserEnabled) {
 			res.setResult(AttackStatus.INVALID);
 			return res;
 		}
+
+		//right now, this only produces one res per attack, even if there are two ships there.
 		this.attacks.add(res);
+
+		//
 		if(shipPresent(x,y)){
-			res.setShip(getShip(x,y));
-
-			if (!isCaptainsQuarter(x,y) ) {
-				res.setResult(AttackStatus.HIT);
-			} else if (isCaptainsQuarter(x,y) && res.getShip().isSunk()) {
-				res.setResult(AttackStatus.SUNK);
-				remainingShips--;
-				if (!this.shipsLeft()) {
-					res.setResult(AttackStatus.SURRENDER);
+			//we'll get all the ships at this location
+			for(Ship ships : getShip(x,y)) {
```

Comment:
For starters, Nicely done. If we had refactored the board class to hold a grid inside of itself instead of giving the ships the squares, it probably would have made this part a bit easier. I ran into the same issue when trying to move the ships.